File: coding-labs/bounded-producer-consumer/java/src/main/java/BoundedQueue.java
import java.time.Duration;
import java.util.ArrayDeque;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

public final class BoundedQueue<T> {
  public enum Status { ITEM, TIMEOUT, CLOSED }
  public record PollResult<T>(Status status, T item) {}
  private final ArrayDeque<T> items = new ArrayDeque<>(); private final int capacity; private final ReentrantLock lock = new ReentrantLock(); private final Condition notEmpty = lock.newCondition(); private final Condition notFull = lock.newCondition(); private boolean closed;
  public BoundedQueue(int capacity) { if (capacity <= 0) throw new IllegalArgumentException("capacity must be positive"); this.capacity = capacity; }
  public boolean offer(T item, Duration timeout) throws InterruptedException {
    long remaining = timeout.toNanos(); lock.lockInterruptibly();
    try { while (!closed && items.size() == capacity) { if (remaining <= 0) return false; remaining = notFull.awaitNanos(remaining); } if (closed) return false; items.addLast(item); notEmpty.signal(); return true; } finally { lock.unlock(); }
  }
  public PollResult<T> poll(Duration timeout) throws InterruptedException {
    long remaining = timeout.toNanos(); lock.lockInterruptibly();
    try { while (items.isEmpty() && !closed) { if (remaining <= 0) return new PollResult<>(Status.TIMEOUT, null); remaining = notEmpty.awaitNanos(remaining); } if (!items.isEmpty()) { T item = items.removeFirst(); notFull.signal(); return new PollResult<>(Status.ITEM, item); } return new PollResult<>(Status.CLOSED, null); } finally { lock.unlock(); }
  }
  public void close() { lock.lock(); try { closed = true; notEmpty.signalAll(); notFull.signalAll(); } finally { lock.unlock(); } }
}

File: coding-labs/bounded-producer-consumer/python/src/bounded_queue.py
from collections import deque
from threading import Condition
from time import monotonic

class BoundedQueue:
    def __init__(self, capacity, on_wait=None):
        if capacity <= 0: raise ValueError("capacity must be positive")
        self._items, self._capacity, self._closed, self._condition = deque(), capacity, False, Condition()
        self._on_wait = on_wait or (lambda: None)

    def offer(self, item, timeout):
        deadline = monotonic() + timeout
        with self._condition:
            while not self._closed and len(self._items) == self._capacity:
                remaining = deadline - monotonic()
                if remaining <= 0: return False
                self._on_wait(); self._condition.wait(remaining)
            if self._closed: return False
            self._items.append(item); self._condition.notify(); return True

    def poll(self, timeout):
        deadline = monotonic() + timeout
        with self._condition:
            while not self._items and not self._closed:
                remaining = deadline - monotonic()
                if remaining <= 0: return ("timeout", None)
                self._on_wait(); self._condition.wait(remaining)
            if self._items:
                item = self._items.popleft(); self._condition.notify(); return ("item", item)
            return ("closed", None)

    def close(self):
        with self._condition:
            self._closed = True; self._condition.notify_all()
